File: packages/oxlint-plugin-react-doctor/src/plugin/rules/a11y/shadcn-dialog-content-requires-title.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getTrailingJsxNameSegment } from "../../utils/get-trailing-jsx-name-segment.js";
import { hasJsxPropIgnoreCase } from "../../utils/has-jsx-prop-ignore-case.js";
import { hasJsxSpreadAttribute } from "../../utils/has-jsx-spread-attribute.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import {
  SHADCN_UI_MODULE_SOURCE_PATTERN,
  resolveShadcnUiComponentName,
} from "../../utils/resolve-shadcn-ui-component-name.js";
import type { RuleContext } from "../../utils/rule-context.js";
import type { JsxSubtreePartScan } from "../../utils/scan-jsx-subtree-for-part.js";
import { scanJsxSubtreeForPart } from "../../utils/scan-jsx-subtree-for-part.js";

interface DialogSurfaceContract {
  readonly contentComponent: string;
  readonly titleComponent: string;
  readonly moduleSourcePattern: RegExp;
}

// Radix's Dialog primitive (which shadcn's dialog, sheet, alert-dialog, and
// vaul-backed drawer modules wrap) requires a Title inside Content — without
// one the dialog has no accessible name and Radix logs a runtime
// accessibility error.
const DIALOG_SURFACE_CONTRACTS: ReadonlyArray<DialogSurfaceContract> = [
  {
    contentComponent: "DialogContent",
    titleComponent: "DialogTitle",
    moduleSourcePattern: /(?:^|\/)ui\/(?:.*\/)?dialog$|^\.\.?\/(?:.*\/)?dialog$/,
  },
  {
    contentComponent: "SheetContent",
    titleComponent: "SheetTitle",
    moduleSourcePattern: /(?:^|\/)ui\/(?:.*\/)?sheet$|^\.\.?\/(?:.*\/)?sheet$/,
  },
  {
    contentComponent: "AlertDialogContent",
    titleComponent: "AlertDialogTitle",
    moduleSourcePattern: /(?:^|\/)ui\/(?:.*\/)?alert-dialog$|^\.\.?\/(?:.*\/)?alert-dialog$/,
  },
  {
    contentComponent: "DrawerContent",
    titleComponent: "DrawerTitle",
    moduleSourcePattern: /(?:^|\/)ui\/(?:.*\/)?drawer$|^\.\.?\/(?:.*\/)?drawer$/,
  },
];

// `aria-label` names the dialog directly; `aria-labelledby` delegates to any
// element; `title` is the accessible-name fallback. Presence (even dynamic)
// counts as named.
const NAME_PROVIDING_ATTRIBUTES = ["aria-label", "aria-labelledby", "title"] as const;

const isTitleElementName = (
  elementName: EsTreeNode,
  contract: DialogSurfaceContract,
  context: RuleContext,
): boolean =>
  resolveShadcnUiComponentName(elementName, contract.moduleSourcePattern, context) ===
    contract.titleComponent ||
  // A local wrapper or re-export named exactly like the title part —
  // trusting the name here trades a rare false negative for zero noise on
  // projects that centralize their dialog headers.
  getTrailingJsxNameSegment(elementName) === contract.titleComponent;

const scanContentForTitle = (
  contentElement: EsTreeNodeOfType<"JSXElement">,
  contract: DialogSurfaceContract,
  context: RuleContext,
): JsxSubtreePartScan =>
  scanJsxSubtreeForPart(contentElement.children, {
    isPartElementName: (elementName) => isTitleElementName(elementName, contract, context),
    // An unresolved custom component may render the title itself
    // (`<ConfirmDialogHeader/>`), so a missing-title claim is unprovable;
    // components generated into `ui/` modules are known leaves. Opaque
    // elements still recurse, so a title nested through them counts.
    isOpaqueElement: (element) => {
      const elementName = element.openingElement.name;
      // Some shadcn-compatible distributions render the header part's text
      // AS the dialog title (Intent UI's DrawerHeader), so a same-module
      // header with content makes the missing-title claim unprovable.
      const resolvedPartName = resolveShadcnUiComponentName(
        elementName,
        contract.moduleSourcePattern,
        context,
      );
      if (resolvedPartName !== null) {
        return (
          resolvedPartName === "DrawerHeader" &&
          element.children.some(
            (child) => isNodeOfType(child, "JSXText") && child.value.trim().length > 0,
          )
        );
      }
      const trailingSegment = getTrailingJsxNameSegment(elementName);
      return (
        trailingSegment !== null &&
        /^[A-Z]/.test(trailingSegment) &&
        trailingSegment !== "Fragment" &&
        resolveShadcnUiComponentName(elementName, SHADCN_UI_MODULE_SOURCE_PATTERN, context) === null
      );
    },
  });

export const shadcnDialogContentRequiresTitle = defineRule({
  id: "shadcn-dialog-content-requires-title",
  title: "Dialog content without a title",
  severity: "warn",
  requires: ["shadcn"],
  recommendation:
    "Give every DialogContent, SheetContent, AlertDialogContent, and DrawerContent a matching title part (wrapped in an sr-only element when the design shows no heading) or name the dialog with aria-label.",
  create: (context: RuleContext) => ({
    JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
      for (const contract of DIALOG_SURFACE_CONTRACTS) {
        if (
          resolveShadcnUiComponentName(node.name, contract.moduleSourcePattern, context) !==
          contract.contentComponent
        ) {
          continue;
        }
        // A spread can supply aria-label / children at runtime.
        if (hasJsxSpreadAttribute(node.attributes)) return;
        if (
          NAME_PROVIDING_ATTRIBUTES.some((attribute) =>
            hasJsxPropIgnoreCase(node.attributes, attribute),
          )
        ) {
          return;
        }
        const element = node.parent;
        if (!element || !isNodeOfType(element, "JSXElement") || element.children.length === 0) {
          return;
        }
        const scan = scanContentForTitle(element, contract, context);
        if (scan.foundPart || scan.sawOpaqueContent) return;
        context.report({
          node: node.name,
          message: `This ${contract.contentComponent} renders no ${contract.titleComponent}, so the dialog has no accessible name and assistive technology announces an unnamed dialog. Add a ${contract.titleComponent} (visually hidden if the design shows no heading) or an aria-label.`,
        });
        return;
      }
    },
  }),
});

File: native/oxlint/rules/shadcn-dialog-content-requires-title.rs
use oxc_ast::{
    ast::{JSXAttributeItem, JSXChild},
    AstKind,
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;

use crate::{context::LintContext, rule::Rule, utils::has_jsx_prop_ignore_case, AstNode};

const DIALOG_MESSAGE: &str = "This DialogContent renders no DialogTitle, so the dialog has no accessible name and assistive technology announces an unnamed dialog. Add a DialogTitle (visually hidden if the design shows no heading) or an aria-label.";
const SHEET_MESSAGE: &str = "This SheetContent renders no SheetTitle, so the dialog has no accessible name and assistive technology announces an unnamed dialog. Add a SheetTitle (visually hidden if the design shows no heading) or an aria-label.";
const ALERT_DIALOG_MESSAGE: &str = "This AlertDialogContent renders no AlertDialogTitle, so the dialog has no accessible name and assistive technology announces an unnamed dialog. Add a AlertDialogTitle (visually hidden if the design shows no heading) or an aria-label.";
const DRAWER_MESSAGE: &str = "This DrawerContent renders no DrawerTitle, so the dialog has no accessible name and assistive technology announces an unnamed dialog. Add a DrawerTitle (visually hidden if the design shows no heading) or an aria-label.";
const NAME_PROVIDING_ATTRIBUTES: [&str; 3] = ["aria-label", "aria-labelledby", "title"];
const DIALOG_SURFACES: [(&str, &str, &str); 4] = [
    ("dialog", "DialogContent", "DialogTitle"),
    ("sheet", "SheetContent", "SheetTitle"),
    ("alert-dialog", "AlertDialogContent", "AlertDialogTitle"),
    ("drawer", "DrawerContent", "DrawerTitle"),
];

#[derive(Debug, Default, Clone)]
pub struct ShadcnDialogContentRequiresTitle;

declare_oxc_lint!(
    /// Require accessible names on shadcn dialog content.
    ShadcnDialogContentRequiresTitle,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Require accessible names on shadcn dialog content.",
);

impl Rule for ShadcnDialogContentRequiresTitle {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        if !has_capability_or_unspecified(ctx, "shadcn") {
            return;
        }
        let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
            return;
        };
        for (module_name, content_component, title_component) in DIALOG_SURFACES {
            if resolve_shadcn_component_name(&opening_element.name, module_name, ctx).as_deref()
                != Some(content_component)
            {
                continue;
            }
            if opening_element
                .attributes
                .iter()
                .any(|attribute| matches!(attribute, JSXAttributeItem::SpreadAttribute(_)))
                || NAME_PROVIDING_ATTRIBUTES.iter().any(|attribute_name| {
                    has_jsx_prop_ignore_case(opening_element, attribute_name).is_some()
                })
            {
                return;
            }
            let parent = ctx.nodes().parent_node(node.id());
            let AstKind::JSXElement(element) = parent.kind() else {
                return;
            };
            if element.children.is_empty() {
                return;
            }
            let scan = scan_static_jsx_subtree_for_part(
                &element.children,
                ctx,
                |child_name| {
                    resolve_shadcn_component_name(child_name, module_name, ctx).as_deref()
                        == Some(title_component)
                        || jsx_element_name_trailing_segment(child_name) == Some(title_component)
                },
                |child_element| {
                    let child_name = &child_element.opening_element.name;
                    if let Some(resolved_part) =
                        resolve_shadcn_component_name(child_name, module_name, ctx)
                    {
                        return resolved_part == "DrawerHeader"
                            && child_element.children.iter().any(|child| {
                                matches!(child, JSXChild::Text(text) if !text.value.trim().is_empty())
                            });
                    }
                    jsx_element_name_trailing_segment(child_name).is_some_and(|name| {
                        name != "Fragment"
                            && name.as_bytes().first().is_some_and(u8::is_ascii_uppercase)
                            && resolve_general_shadcn_ui_component_name(child_name, ctx).is_none()
                    })
                },
            );
            if scan.found_part || scan.saw_opaque_content {
                return;
            }
            let message = match module_name {
                "dialog" => DIALOG_MESSAGE,
                "sheet" => SHEET_MESSAGE,
                "alert-dialog" => ALERT_DIALOG_MESSAGE,
                "drawer" => DRAWER_MESSAGE,
                _ => return,
            };
            ctx.diagnostic(OxcDiagnostic::warn(message).with_label(opening_element.name.span()));
            return;
        }
    }
}

fn resolve_shadcn_component_name<'a>(
    element_name: &oxc_ast::ast::JSXElementName<'a>,
    module_name: &str,
    ctx: &LintContext<'a>,
) -> Option<String> {
    let api_path = resolve_jsx_import_api_path(
        element_name,
        |module_source| is_shadcn_component_module(module_source, module_name),
        ctx,
    )?;
    let [component_name] = api_path.as_slice() else {
        return None;
    };
    Some(component_name.clone())
}

fn resolve_general_shadcn_ui_component_name<'a>(
    element_name: &oxc_ast::ast::JSXElementName<'a>,
    ctx: &LintContext<'a>,
) -> Option<String> {
    let api_path = resolve_jsx_import_api_path(element_name, is_general_shadcn_ui_module, ctx)?;
    let [component_name] = api_path.as_slice() else {
        return None;
    };
    Some(component_name.clone())
}

fn is_shadcn_component_module(module_source: &str, module_name: &str) -> bool {
    let normalized_source = module_source.replace('\\', "/");
    let expected_suffix = format!("/{module_name}");
    normalized_source.ends_with(&expected_suffix)
        && (normalized_source.starts_with("./")
            || normalized_source.starts_with("../")
            || normalized_source == format!("ui/{module_name}")
            || normalized_source.contains("/ui/"))
}

fn is_general_shadcn_ui_module(module_source: &str) -> bool {
    let normalized_source = module_source.replace('\\', "/");
    let Some(ui_tail) = normalized_source.strip_prefix("ui/").or_else(|| {
        normalized_source
            .split_once("/ui/")
            .map(|(_, ui_tail)| ui_tail)
    }) else {
        return false;
    };
    let component_tail = ui_tail.strip_prefix("components/").unwrap_or(ui_tail);
    !component_tail.is_empty() && !component_tail.contains('/')
}
